### PR TITLE
Custom middleware

### PR DESCRIPTION
**Pull request by kvigen**
This change adds support for specifying custom middleware. We've had enough
people ask for it that we thought it was worth coming up with a plan.

- This change pulls all the middleware code out into a separate wag package so that it isn't autogenerated.
- It adds a `NewWithMiddleware` method to server creation

I'm going to follow this up with a change to add mongo middleware so we can try that out

### CODE REVIEW

**Comment by kvigen** [comment]
cc @Christopher-Bradshaw

**Comment by Christopher-Bradshaw** [comment]
Damn, so fast 👍

**Comment by cozmo** [comment]
@kvigen Can you separate your changes and the generated code in order? Like first all your commits and then a final commit with generated code? Would make this a lot easier to review for me.

**Comment by kvigen** [comment]
@cozmo put all the generated code in the "Generated Code" commit

File: README.md [hunk 1]
```diff
@@ -67,6 +67,13 @@ To use the generated code you need to do two things:
   // Serve should not return
   log.Fatal(s.Serve())
 ```
+Or if you want to add custom middleware
+```
+  s := server.New(myController, ":8000", []func(http.Handler) http.Handler{
```
Comment (cozmo) [inline README.md]: I think you're missing a `NewWithMiddleware ` call here?

Comment (kvigen) [inline README.md]: derp

**Pull request merged**